upgrade/v121to122/upgrade.go

```diff
@@ -0,0 +1,140 @@
+package v121to122
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/longhorn/longhorn-manager/engineapi"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1.2.1 to v1.2.2: "
+)
+
+func UpgradeCRs(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"UpgradeCRs failed")
+	}()
+	if err := upgradeBackups(namespace, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeEngines(namespace, lhClient); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeBackups(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "upgrade backups failed")
+	}()
+
+	// Copy backupStatus from engine CRs to backup CRs
+	backups, err := lhClient.LonghornV1beta1().Backups(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Loop all the backup CRs
+	for _, backup := range backups.Items {
+		// Get volume name from label
+		volumeName, exist := backup.Labels[types.LonghornLabelBackupVolume]
+		if !exist {
+			continue
+		}
+
+		// Get the corresponding volume's engine
+		engines, err := lhClient.LonghornV1beta1().Engines(namespace).List(context.TODO(), metav1.ListOptions{
+			LabelSelector: types.LonghornLabelVolume + "=" + volumeName,
+		})
+		if err != nil {
+			continue
+		}
+
+		// No engine CR found
+		var engine longhorn.Engine
+		switch len(engines.Items) {
+		case 0:
+			continue
+		case 1:
+			engine = engines.Items[0]
+		default:
+			// If more than 2 engines found, use the current volume's engine
+			v, err := lhClient.LonghornV1beta1().Volumes(namespace).Get(context.TODO(), volumeName, metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					// Cannot found the correspoding volume
+					continue
+				}
+				return err
```

Comment:
This will cause upgrade failed in the middle? If yes, I am thinking if we can make some pre-upgrade validation in the future in general to make sure at least the upgrade condition is OK before the upgrade (not really related to this PR)

Reply:
I think no. It's because the Backup CR doesn't have a corresponding Volume within the cluster (Backup pulled from the remote backup target)